go/login/server/keymanager.go

```diff
@@ -0,0 +1,164 @@
+type ErrInvalidKeyIndex struct {
+	Index uint8
+}
+
+func (e ErrInvalidKeyIndex) Error() string {
+	return fmt.Sprintf("key index should be in range 0-127, found: %v", e.Index)
+}
+
+// ErrOverloadedKeyIndex denotes that provided index is already in use, so no
+// new keys can be assigned to it.
+type ErrOverloadedKeyIndex struct {
+	Index uint8
+}
+
+func (e ErrOverloadedKeyIndex) Error() string {
+	return fmt.Sprintf("key index already in use, found: %v", e.Index)
+}
+
+// ErrKeyIndexNotFound denotes that an index was not found, so no key can be
+// returned from it.
+type ErrKeyIndexNotFound struct {
+	Index uint8
+}
+
+func (e ErrKeyIndexNotFound) Error() string {
+	return fmt.Sprintf("key index %v not found", e.Index)
+}
+
+// A PrivateKey represent a Private key for the login server. It is a pair composed of a private key
+// and its pairing index.
+type PrivateKey struct {
+	Value glome.PrivateKey
+	Index uint8
+}
+
+// A PublicKey represent a Service key for the login server. It is a pair composed of a public key
+// and its pairing index.
+type PublicKey struct {
+	Value glome.PublicKey
+	Index uint8
+}
+
+// KeyManager performs key maneger task in a concurrent-safe way. It allows for constant
+// time search of keys by index. KeyManager is constructed with NewKeyManager function.
+type KeyManager struct {
+	indexToPriv map[uint8]glome.PrivateKey
+	publicKeys  []PublicKey
+	lock        sync.RWMutex
+}
+
+func (k *KeyManager) locklessAdd(key glome.PrivateKey, index uint8) error {
+	if index > 127 {
+		return ErrInvalidKeyIndex{Index: index}
+	}
+	if _, ok := k.indexToPriv[index]; ok {
+		return ErrOverloadedKeyIndex{Index: index}
+	}
+
+	pub, err := key.Public()
+	if err != nil {
+		return err
+	}
+
+	k.indexToPriv[index] = key
+	k.publicKeys = append(k.publicKeys, PublicKey{Value: *pub, Index: index})
+	return nil
+}
+
+// Add adds provided key and index to the key manager. Raises ErrInvalidindex
+// if index provided is not in range 0-127 and ErrOverloadedIndex if index
+// provided was already in use.
+func (k *KeyManager) Add(key glome.PrivateKey, index uint8) error {
+	k.lock.Lock()
+	defer k.lock.Unlock()
+	return k.locklessAdd(key, index)
+}
+
+// Read returns the PrivateKey stored in the KeyManager for a index, or a
+// zero-value PrivateKey if no PrivateKey is present. The ok result indicates
+// whether value was found in the KeyManager.
+func (k *KeyManager) Read(index uint8) (glome.PrivateKey, bool) {
+	k.lock.RLock()
+	defer k.lock.RUnlock()
+
+	key, ok := k.indexToPriv[index]
+	return key, ok
+}
+
+// DropAllReplace drops all stored keys and replace them with the new ones provided.
+// This operation is done in a atomic way (no other call to the struct will be handled
+// while DropAllReplace is).
+func (k *KeyManager) DropAllReplace(keys []PrivateKey) error {
+	k.lock.Lock()
+	defer k.lock.Unlock()
+
+	k.indexToPriv = make(map[uint8]glome.PrivateKey)
+	k.publicKeys = make([]PublicKey, 0)
```

Comment:
No need to make() this. You can just assign `nil` and `append()` will DTRT.

Reply:
Deleted.